File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.js
/**
 * @param {number} m
 * @param {number} n
 * @param {number[][]} guards
 * @param {number[][]} walls
 * @return {number}
 */
var countUnguarded = function (m, n, guards, walls) {
    let grid = Array.from({ length: m }, () => Array(n).fill(0));

    // Mark guards and walls as 2
    for (let [x, y] of [...guards, ...walls]) {
        grid[x][y] = 2;
    }


    // Mark guarded cells
    for (let [i, j] of guards) {
        for (let [x, y] of [[-1, 0], [0, 1], [1, 0], [0, -1]]) {  // Directions: up, left, down, right
            let a = i;
            let b = j;
            while (true) {
                a += x;
                b += y;
                // Check boundaries and stop at walls or other guards
                if (a < 0 || a >= m || b < 0 || b >= n || grid[a][b] === 2) {
                    break;
                }
                grid[a][b] = 1;  // Mark the cell as guarded
            }
        }
    }

    // Count unguarded cells
    let count = 0
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            if (grid[i][j] == 0) {
                count++
            }
        }
    }
    return count


}

File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.py
class Solution:
    def countUnguarded(self, m: int, n: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        grid = [[0] * n  for _ in range(m)]
        
        for x, y in guards + walls:
            grid[x][y] = 2
        for di, dj in guards:
            for dx, dy in [[-1, 0], [0, -1], [1, 0], [0, 1]]:
                x = di
                y = dj
                while True:
                    x += dx
                    y += dy
                    if x < 0 or y < 0 or x >= m or y >=n or grid[x][y] == 2:
                        break
                    grid[x][y] = 1
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 0:
                    count += 1
        return count
